api/pom.xml

```diff
@@ -43,6 +43,19 @@
             <artifactId>openmrs-test</artifactId>
             <type>pom</type>
         </dependency>
+        <dependency>
```

Comment:
To keep this simple, I think conversion of the tests to use testcontainers should be done separately as part of another ticket, just run H2 in mysql mode for now

Reply:
Hi @wluyima, I've tried the option you are talking about and the error never occurs. It's as if the option `MODE=MYSQL` is never considered altogether.

Reply:
How are you doing it? Do you mind sharing the code snippet?

Reply:
I always make tests fail on H2 that pass on MySQL by setting the mode and it is why I ended up adding [this](https://github.com/openmrs/openmrs-module-datafilter/blob/2.0.0/api/src/main/java/org/openmrs/module/datafilter/impl/ImplDataFilterListener.java#L99) line and [this](https://github.com/openmrs/openmrs-module-datafilter/blob/2.0.0/api/src/main/java/org/openmrs/module/datafilter/impl/ImplDataFilterListener.java#L105), if you comment them out and set mode to mysql, some tests will fail and the same should happen for you, unless you are missing something.

Reply:
@wluyima, I'm doing the above to reproduce the exact runtime error. Otherwise a different non-specific error occurs. See snippet in draft [here](https://github.com/openmrs/openmrs-module-datafilter/pull/21).

Reply:
This is the error that shows up.
```
Syntax error in SQL statement "SELECT THIS_.PROVIDER_ID AS PROVIDER1_87_0_, THIS_.PERSON_ID AS PERSON_I2_87_0_, THIS_.NAME AS NAME3_87_0_, THIS_.IDENTIFIER AS IDENTIFI4_87_0_, THIS_.CREATOR AS CREATOR5_87_0_, THIS_.ROLE_ID AS ROLE_ID6_87_0_, THIS_.SPECIALITY_ID AS SPECIALI7_87_0_, THIS_.DATE_CREATED AS DATE_CRE8_87_0_, THIS_.CHANGED_BY AS CHANGED_9_87_0_, THIS_.DATE_CHANGED AS DATE_CH10_87_0_, THIS_.RETIRED_BY AS RETIRED11_87_0_, THIS_.DATE_RETIRED AS DATE_RE12_87_0_, THIS_.RETIRE_REASON AS RETIRE_13_87_0_, THIS_.RETIRED AS RETIRED14_87_0_, THIS_.UUID AS UUID15_87_0_ FROM PROVIDER THIS_ WHERE (
            THIS_.PERSON_ID IN (
                SELECT DATAFILTER_US.PERSON_ID FROM USERS DATAFILTER_US
                    WHERE DATAFILTER_US.USER_ID IN (
                        SELECT DISTINCT DATAFILTER_UR.USER_ID FROM USER_ROLE DATAFILTER_UR
                            WHERE DATAFILTER_UR.ROLE IN ()[*]
                            OR DATAFILTER_UR.ROLE NOT IN ()
                    )
                    OR DATAFILTER_US.USER_ID NOT IN (SELECT DISTINCT DATAFILTER_UR2.USER_ID FROM USER_ROLE DATAFILTER_UR2)
                    OR DATAFILTER_US.USER_ID IN (
                        SELECT DISTINCT DATAFILTER_UR1.USER_ID FROM USER_ROLE DATAFILTER_UR1
                            WHERE DATAFILTER_UR1.ROLE IN (
                                SELECT DISTINCT DATAFILTER_EBM.ENTITY_IDENTIFIER FROM DATAFILTER_ENTITY_BASIS_MAP DATAFILTER_EBM
                                    WHERE DATAFILTER_EBM.ENTITY_TYPE = 'org.openmrs.Role'
                                    AND DATAFILTER_EBM.BASIS_TYPE = 'org.openmrs.Program'
                                    AND DATAFILTER_EBM.BASIS_IDENTIFIER IN (
                                        SELECT DISTINCT DATAFILTER_EBM2.BASIS_IDENTIFIER FROM DATAFILTER_ENTITY_BASIS_MAP DATAFILTER_EBM2
                                            WHERE DATAFILTER_EBM2.ENTITY_TYPE = 'org.openmrs.Role'
                                                AND DATAFILTER_EBM2.BASIS_TYPE = 'org.openmrs.Program'
                                                AND DATAFILTER_EBM2.ENTITY_IDENTIFIER IN ()
                                    )
                            )
                    )
            )
        ) AND THIS_.PERSON_ID=? ORDER BY THIS_.RETIRED ASC, THIS_.PROVIDER_ID ASC"; expected "NOT, EXISTS, INTERSECTS, UNIQUE"; SQL statement:
select this_.provider_id as provider1_87_0_, this_.person_id as person_i2_87_0_, this_.name as name3_87_0_, this_.identifier as identifi4_87_0_, this_.creator as creator5_87_0_, this_.role_id as role_id6_87_0_, this_.speciality_id as speciali7_87_0_, this_.date_created as date_cre8_87_0_, this_.changed_by as changed_9_87_0_, this_.date_changed as date_ch10_87_0_, this_.retired_by as retired11_87_0_, this_.date_retired as date_re12_87_0_, this_.retire_reason as retire_13_87_0_, this_.retired as retired14_87_0_, this_.uuid as uuid15_87_0_ from provider this_ where (
            this_.person_id IN (
                SELECT datafilter_us.person_id FROM users datafilter_us
                    WHERE datafilter_us.user_id IN (
                        SELECT DISTINCT datafilter_ur.user_id FROM user_role datafilter_ur
                            WHERE datafilter_ur.role IN ()
                            OR datafilter_ur.role NOT IN ()
                    )
                    OR datafilter_us.user_id NOT IN (SELECT DISTINCT datafilter_ur2.user_id FROM user_role datafilter_ur2)
                    OR datafilter_us.user_id IN (
                        SELECT DISTINCT datafilter_ur1.user_id FROM user_role datafilter_ur1
                            WHERE datafilter_ur1.role IN (
                                SELECT DISTINCT datafilter_ebm.entity_identifier FROM datafilter_entity_basis_map datafilter_ebm
                                    WHERE datafilter_ebm.entity_type = 'org.openmrs.Role'
                                    AND datafilter_ebm.basis_type = 'org.openmrs.Program'
                                    AND datafilter_ebm.basis_identifier IN (
                                        SELECT DISTINCT datafilter_ebm2.basis_identifier FROM datafilter_entity_basis_map datafilter_ebm2
                                            WHERE datafilter_ebm2.entity_type = 'org.openmrs.Role'
                                                AND datafilter_ebm2.basis_type = 'org.openmrs.Program'
                                                AND datafilter_ebm2.entity_identifier IN ()
                                    )
                            )
                    )
            )
        ) and this_.person_id=? order by this_.retired asc, this_.provider_id asc [42001-200]
```

Reply:
When you do as I said, of course the above error is expected which implies that H2 is correctly running in MySQL mode since MySQL doesn't support the generated query, and when you uncomment the 2 lines I linked to in my earlier comment the tests should pass, so what's the problem then?

Reply:
The problem is, the runtime vs test errors look differently. I can only assume they mean the same thing for the moment. That said I'll go ahead and use the MODE with a generic error for assertion.

Reply:
And what's the difference? Please share the 2 sets of errors

Reply:
Please share the full error stacktrace for both situations(deployed vs test) and not just portions

Reply:
FYI I don't know if you realize that [these lines with ####](https://github.com/openmrs/openmrs-module-datafilter/blob/2.0.0/api/src/main/java/org/openmrs/module/datafilter/impl/ImplDataFilterListener.java#L99) address the same issue you are trying to fix as part of the ticket with the ONLY difference being in your case there is no authenticated user, however with or without an authenticated the resultant root error is the same i.e. a bad query with a blank sql `IN` clause

Reply:
The runtime error seen in this PR is
```
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')
                           OR datafilter_ur.role NOT IN ()
                  ' at line 6"
```

Reply:
And which is the other one?

Reply:
I shared the other one [above](https://github.com/openmrs/openmrs-module-datafilter/pull/20#discussion_r1060146621).

Reply:
Please share full stacktrace, that is just a portion

Reply:
Runtime/test failure with actual MySQL env

```
org.hibernate.exception.SQLGrammarException: could not extract ResultSet
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:63)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:67)
	at org.hibernate.loader.Loader.getResultSet(Loader.java:2341)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2094)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2056)
	at org.hibernate.loader.Loader.doQuery(Loader.java:953)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:350)
	at org.hibernate.loader.Loader.doList(Loader.java:2887)
	at org.hibernate.loader.Loader.doList(Loader.java:2869)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2701)
	at org.hibernate.loader.Loader.list(Loader.java:2696)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1878)
	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:370)
	at org.openmrs.api.db.hibernate.HibernateProviderDAO.getProvidersByPerson(HibernateProviderDAO.java:113)
	at org.openmrs.api.impl.ProviderServiceImpl.getProvidersByPerson(ProviderServiceImpl.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:367)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy183.getProvidersByPerson(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.cache.interceptor.CacheInterceptor.lambda$invoke$0(CacheInterceptor.java:53)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:352)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:123)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy184.getProvidersByPerson(Unknown Source)
	at org.openmrs.api.impl.ProviderServiceImpl.getProvidersByPerson(ProviderServiceImpl.java:137)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:367)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy183.getProvidersByPerson(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.cache.interceptor.CacheInterceptor.lambda$invoke$0(CacheInterceptor.java:53)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:352)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:123)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy184.getProvidersByPerson(Unknown Source)
	at org.openmrs.module.datafilter.impl.UserProgramBasedFilterTest.getProvidersByPerson_shouldFailWithSQLSyntaxErrorException(UserProgramBasedFilterTest.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:43)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:82)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:73)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:229)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:197)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:191)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:137)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.sql.SQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')
                            OR datafilter_ur.role NOT IN ()
                  ' at line 6
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1003)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:57)
	... 134 more
```


Error with H2 in MODE=MYSQL

```
org.hibernate.exception.SQLGrammarException: could not prepare statement
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:63)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:186)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:151)
	at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:2141)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2078)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2056)
	at org.hibernate.loader.Loader.doQuery(Loader.java:953)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:350)
	at org.hibernate.loader.Loader.doList(Loader.java:2887)
	at org.hibernate.loader.Loader.doList(Loader.java:2869)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2701)
	at org.hibernate.loader.Loader.list(Loader.java:2696)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1878)
	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:370)
	at org.openmrs.api.db.hibernate.HibernateProviderDAO.getProvidersByPerson(HibernateProviderDAO.java:113)
	at org.openmrs.api.impl.ProviderServiceImpl.getProvidersByPerson(ProviderServiceImpl.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:367)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy174.getProvidersByPerson(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.cache.interceptor.CacheInterceptor.lambda$invoke$0(CacheInterceptor.java:53)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:352)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:123)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy175.getProvidersByPerson(Unknown Source)
	at org.openmrs.api.impl.ProviderServiceImpl.getProvidersByPerson(ProviderServiceImpl.java:137)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:367)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy174.getProvidersByPerson(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.cache.interceptor.CacheInterceptor.lambda$invoke$0(CacheInterceptor.java:53)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:352)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:123)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy175.getProvidersByPerson(Unknown Source)
	at org.openmrs.module.datafilter.impl.UserProgramBasedFilterTest.getProvidersByPerson_shouldFailWithSQLSyntaxErrorException(UserProgramBasedFilterTest.java:133)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:43)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:82)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:73)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:229)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:197)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:191)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:137)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Syntax error in SQL statement "SELECT THIS_.PROVIDER_ID AS PROVIDER1_87_0_, THIS_.PERSON_ID AS PERSON_I2_87_0_, THIS_.NAME AS NAME3_87_0_, THIS_.IDENTIFIER AS IDENTIFI4_87_0_, THIS_.CREATOR AS CREATOR5_87_0_, THIS_.ROLE_ID AS ROLE_ID6_87_0_, THIS_.SPECIALITY_ID AS SPECIALI7_87_0_, THIS_.DATE_CREATED AS DATE_CRE8_87_0_, THIS_.CHANGED_BY AS CHANGED_9_87_0_, THIS_.DATE_CHANGED AS DATE_CH10_87_0_, THIS_.RETIRED_BY AS RETIRED11_87_0_, THIS_.DATE_RETIRED AS DATE_RE12_87_0_, THIS_.RETIRE_REASON AS RETIRE_13_87_0_, THIS_.RETIRED AS RETIRED14_87_0_, THIS_.UUID AS UUID15_87_0_ FROM PROVIDER THIS_ WHERE (
            THIS_.PERSON_ID IN (
                SELECT DATAFILTER_US.PERSON_ID FROM USERS DATAFILTER_US
                    WHERE DATAFILTER_US.USER_ID IN (
                        SELECT DISTINCT DATAFILTER_UR.USER_ID FROM USER_ROLE DATAFILTER_UR
                            WHERE DATAFILTER_UR.ROLE IN ()[*]
                            OR DATAFILTER_UR.ROLE NOT IN ()
                    )
                    OR DATAFILTER_US.USER_ID NOT IN (SELECT DISTINCT DATAFILTER_UR2.USER_ID FROM USER_ROLE DATAFILTER_UR2)
                    OR DATAFILTER_US.USER_ID IN (
                        SELECT DISTINCT DATAFILTER_UR1.USER_ID FROM USER_ROLE DATAFILTER_UR1
                            WHERE DATAFILTER_UR1.ROLE IN (
                                SELECT DISTINCT DATAFILTER_EBM.ENTITY_IDENTIFIER FROM DATAFILTER_ENTITY_BASIS_MAP DATAFILTER_EBM
                                    WHERE DATAFILTER_EBM.ENTITY_TYPE = 'org.openmrs.Role'
                                    AND DATAFILTER_EBM.BASIS_TYPE = 'org.openmrs.Program'
                                    AND DATAFILTER_EBM.BASIS_IDENTIFIER IN (
                                        SELECT DISTINCT DATAFILTER_EBM2.BASIS_IDENTIFIER FROM DATAFILTER_ENTITY_BASIS_MAP DATAFILTER_EBM2
                                            WHERE DATAFILTER_EBM2.ENTITY_TYPE = 'org.openmrs.Role'
                                                AND DATAFILTER_EBM2.BASIS_TYPE = 'org.openmrs.Program'
                                                AND DATAFILTER_EBM2.ENTITY_IDENTIFIER IN ()
                                    )
                            )
                    )
            )
        ) AND THIS_.PERSON_ID=? ORDER BY THIS_.RETIRED ASC, THIS_.PROVIDER_ID ASC"; expected "NOT, EXISTS, INTERSECTS, UNIQUE"; SQL statement:
select this_.provider_id as provider1_87_0_, this_.person_id as person_i2_87_0_, this_.name as name3_87_0_, this_.identifier as identifi4_87_0_, this_.creator as creator5_87_0_, this_.role_id as role_id6_87_0_, this_.speciality_id as speciali7_87_0_, this_.date_created as date_cre8_87_0_, this_.changed_by as changed_9_87_0_, this_.date_changed as date_ch10_87_0_, this_.retired_by as retired11_87_0_, this_.date_retired as date_re12_87_0_, this_.retire_reason as retire_13_87_0_, this_.retired as retired14_87_0_, this_.uuid as uuid15_87_0_ from provider this_ where (
            this_.person_id IN (
                SELECT datafilter_us.person_id FROM users datafilter_us
                    WHERE datafilter_us.user_id IN (
                        SELECT DISTINCT datafilter_ur.user_id FROM user_role datafilter_ur
                            WHERE datafilter_ur.role IN ()
                            OR datafilter_ur.role NOT IN ()
                    )
                    OR datafilter_us.user_id NOT IN (SELECT DISTINCT datafilter_ur2.user_id FROM user_role datafilter_ur2)
                    OR datafilter_us.user_id IN (
                        SELECT DISTINCT datafilter_ur1.user_id FROM user_role datafilter_ur1
                            WHERE datafilter_ur1.role IN (
                                SELECT DISTINCT datafilter_ebm.entity_identifier FROM datafilter_entity_basis_map datafilter_ebm
                                    WHERE datafilter_ebm.entity_type = 'org.openmrs.Role'
                                    AND datafilter_ebm.basis_type = 'org.openmrs.Program'
                                    AND datafilter_ebm.basis_identifier IN (
                                        SELECT DISTINCT datafilter_ebm2.basis_identifier FROM datafilter_entity_basis_map datafilter_ebm2
                                            WHERE datafilter_ebm2.entity_type = 'org.openmrs.Role'
                                                AND datafilter_ebm2.basis_type = 'org.openmrs.Program'
                                                AND datafilter_ebm2.entity_identifier IN ()
                                    )
                            )
                    )
            )
        ) and this_.person_id=? order by this_.retired asc, this_.provider_id asc [42001-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:453)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.getSyntaxError(DbException.java:243)
	at org.h2.command.Parser.getSyntaxError(Parser.java:1053)
	at org.h2.command.Parser.readTerm(Parser.java:4479)
	at org.h2.command.Parser.readFactor(Parser.java:3343)
	at org.h2.command.Parser.readSum(Parser.java:3330)
	at org.h2.command.Parser.readConcat(Parser.java:3305)
	at org.h2.command.Parser.readCondition(Parser.java:3108)
	at org.h2.command.Parser.readExpression(Parser.java:3059)
	at org.h2.command.Parser.readInPredicate(Parser.java:3274)
	at org.h2.command.Parser.readCondition(Parser.java:3191)
	at org.h2.command.Parser.readExpressionWithGlobalConditions(Parser.java:3046)
	at org.h2.command.Parser.parseSelect(Parser.java:2962)
	at org.h2.command.Parser.parseQuerySub(Parser.java:2817)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:2649)
	at org.h2.command.Parser.parseQuery(Parser.java:2620)
	at org.h2.command.Parser.readInPredicate(Parser.java:3264)
	at org.h2.command.Parser.readCondition(Parser.java:3191)
	at org.h2.command.Parser.readExpressionWithGlobalConditions(Parser.java:3046)
	at org.h2.command.Parser.parseSelect(Parser.java:2962)
	at org.h2.command.Parser.parseQuerySub(Parser.java:2817)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:2649)
	at org.h2.command.Parser.parseQuery(Parser.java:2620)
	at org.h2.command.Parser.readInPredicate(Parser.java:3264)
	at org.h2.command.Parser.readCondition(Parser.java:3191)
	at org.h2.command.Parser.readExpression(Parser.java:3059)
	at org.h2.command.Parser.readTerm(Parser.java:4344)
	at org.h2.command.Parser.readFactor(Parser.java:3343)
	at org.h2.command.Parser.readSum(Parser.java:3330)
	at org.h2.command.Parser.readConcat(Parser.java:3305)
	at org.h2.command.Parser.readCondition(Parser.java:3108)
	at org.h2.command.Parser.readExpressionWithGlobalConditions(Parser.java:3046)
	at org.h2.command.Parser.parseSelect(Parser.java:2962)
	at org.h2.command.Parser.parseQuerySub(Parser.java:2817)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:2649)
	at org.h2.command.Parser.parseQuery(Parser.java:2620)
	at org.h2.command.Parser.parsePrepared(Parser.java:868)
	at org.h2.command.Parser.parse(Parser.java:843)
	at org.h2.command.Parser.parse(Parser.java:819)
	at org.h2.command.Parser.prepareCommand(Parser.java:738)
	at org.h2.engine.Session.prepareLocal(Session.java:657)
	at org.h2.engine.Session.prepareCommand(Session.java:595)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1235)
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:352)
	at com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:567)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:149)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
	... 134 more
```

Reply:
Thanks! The stacktraces might not look 100% identical but if you closely analyze both them, they are reporting the same issue i.e. the `NOT IN` clause having empty braces which is the real issue, H2 seems not to have an issue with it but MySQL does, so using MODE=MySQL should be sufficient to reproduce the bug in tests.

Feel free to create a separate ticket to update the module to use testcontainers